File: engineering/engineering/report/pre_use_report/pre_use_report.py
import frappe
from frappe.utils import getdate


def execute(filters=None):
    if not filters:
        return [], []

    start = getdate(filters.get("start_date"))
    end = getdate(filters.get("end_date"))
    site = filters.get("site")
    shift = filters.get("shift")
    asset_category = filters.get("asset_category")
    asset = filters.get("asset")

    columns = [
        {"label": "Asset Category", "fieldname": "asset_category", "fieldtype": "Data", "width": 140},
        {"label": "Date", "fieldname": "shift_date", "fieldtype": "Date", "width": 120},
        {"label": "Machine No.", "fieldname": "asset_name", "fieldtype": "Data", "width": 130},
        {"label": "Make/Model", "fieldname": "item_name", "fieldtype": "Data", "width": 170},
        {"label": "Start Hours", "fieldname": "start_hours", "fieldtype": "Float", "width": 120},
        {"label": "End Hours", "fieldname": "end_hours", "fieldtype": "Float", "width": 120},
        {"label": "Working Hours", "fieldname": "working_hours", "fieldtype": "Float", "width": 140},
    ]

    conditions = [
        "p.shift_date BETWEEN %(start)s AND %(end)s",
        "p.location = %(site)s"
    ]

    if shift:
        conditions.append("p.shift = %(shift)s")

    if asset_category:
        conditions.append("a.asset_category = %(asset_category)s")

    if asset:
        conditions.append("a.asset_name = %(asset)s")

    query = f"""
        SELECT
            a.asset_category,
            a.asset_name,
            a.item_name,
            p.shift,
            p.shift_date,
            a.eng_hrs_start,
            a.eng_hrs_end,
            a.working_hours
        FROM `tabPre-Use Hours` p
        LEFT JOIN `tabPre-use Assets` a
            ON a.parent = p.name
        WHERE {" AND ".join(conditions)}
        ORDER BY
            a.asset_category,
            a.asset_name,
            p.shift_date,
            p.shift
    """

    rows = frappe.db.sql(query, {
        "start": start,
        "end": end,
        "site": site,
        "shift": shift,
        "asset_category": asset_category,
        "asset": asset
    }, as_dict=True)

    data = []

    if shift:
        for r in rows:
            working_hours = None

            if r.eng_hrs_start is not None and r.eng_hrs_end is not None:
                if float(r.eng_hrs_end) == 0:
                    working_hours = 0
                else:
                    working_hours = round(float(r.eng_hrs_end) - float(r.eng_hrs_start), 1)
            elif r.working_hours is not None:
                working_hours = r.working_hours

            data.append({
                "asset_category": r.asset_category,
                "shift_date": r.shift_date,
                "asset_name": r.asset_name,
                "item_name": r.item_name,
                "start_hours": r.eng_hrs_start,
                "end_hours": r.eng_hrs_end,
                "working_hours": working_hours
            })

    else:
        combo_map = {}

        for r in rows:
            key = (r.asset_category, r.asset_name, r.shift_date)

            if key not in combo_map:
                combo_map[key] = {
                    "asset_category": r.asset_category,
                    "shift_date": r.shift_date,
                    "asset_name": r.asset_name,
                    "item_name": r.item_name,
                    "start_hours": None,
                    "end_hours": None
                }

            if r.shift in ["Day", "Morning"]:
                combo_map[key]["start_hours"] = r.eng_hrs_start
            elif r.shift in ["Night", "Afternoon"]:
                combo_map[key]["end_hours"] = r.eng_hrs_end

        for val in combo_map.values():
            start_hours = val["start_hours"]
            end_hours = val["end_hours"]
            working_hours = None

            if start_hours is not None and end_hours is not None:
                if float(end_hours) == 0:
                    working_hours = 0
                else:
                    working_hours = round(float(end_hours) - float(start_hours), 1)

            data.append({
                "asset_category": val["asset_category"],
                "shift_date": val["shift_date"],
                "asset_name": val["asset_name"],
                "item_name": val["item_name"],
                "start_hours": start_hours,
                "end_hours": end_hours,
                "working_hours": working_hours
            })

    return columns, data

File: engineering/engineering/report/pre_use_report/pre_use_report.js
frappe.query_reports["Pre-Use Report"] = {
    filters: [
        {
            fieldname: "start_date",
            label: __("Start Date"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.add_days(frappe.datetime.get_today(), -7)
        },
        {
            fieldname: "end_date",
            label: __("End Date"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.get_today()
        },
        {
            fieldname: "site",
            label: __("Site"),
            fieldtype: "Link",
            options: "Location",
            reqd: 1
        },
        {
            fieldname: "shift",
            label: __("Shift"),
            fieldtype: "Select",
            options: ["", "Day", "Night", "Morning", "Afternoon"]
        },
        {
            fieldname: "asset_category",
            label: __("Asset Category"),
            fieldtype: "Link",
            options: "Asset Category"
        },
        {
            fieldname: "asset",
            label: __("Machine"),
            fieldtype: "Link",
            options: "Asset"
        }
    ],

    tree: true,
    name_field: "asset_name",
    parent_field: "asset_category",
    initial_depth: 1,

    onload: function(report) {
        if (!$("#pre-use-report-style").length) {
            $("head").append(`
                <style id="pre-use-report-style">
                    .query-report .dt-scrollable {
                        border-radius: 14px;
                        border: 1px solid #dbe3ea;
                        box-shadow: 0 4px 14px rgba(15, 23, 42, 0.08);
                    }

                    .query-report .dt-header {
                        background: #102a43 !important;
                        color: white !important;
                        font-weight: 700 !important;
                    }

                    .query-report .dt-row:nth-child(even) {
                        background: #f8fafc;
                    }

                    .query-report .dt-row:hover {
                        background: #eef6ff !important;
                    }

                    .pre-use-legend {
                        position: absolute;
                        top: 58px;
                        right: 20px;
                        background: #ffffff;
                        border-radius: 14px;
                        padding: 14px 16px;
                        box-shadow: 0 6px 20px rgba(15, 23, 42, 0.18);
                        font-size: 13px;
                        z-index: 20;
                        min-width: 275px;
                        border: 1px solid #e5e7eb;
                    }

                    .pre-use-legend-title {
                        font-size: 14px;
                        font-weight: 800;
                        color: #102a43;
                        margin-bottom: 8px;
                    }

                    .pre-use-pill {
                        padding: 4px 10px;
                        border-radius: 999px;
                        font-weight: 800;
                        display: inline-block;
                        min-width: 62px;
                        text-align: center;
                    }

                    .pre-use-pill-orange {
                        background: #fff3cd;
                        color: #b45309;
                        border: 1px solid #facc15;
                    }

                    .pre-use-pill-red {
                        background: #fee2e2;
                        color: #b91c1c;
                        border: 1px solid #fca5a5;
                    }

                    .pre-use-pill-green {
                        background: #dcfce7;
                        color: #166534;
                        border: 1px solid #86efac;
                    }

                    .pre-use-machine {
                        font-weight: 900;
                        color: #102a43;
                    }

                    .pre-use-muted {
                        color: #64748b;
                        font-weight: 500;
                    }

                    .pre-use-number {
                        font-weight: 700;
                        color: #334155;
                    }
                </style>
            `);
        }

        if (!report.page.wrapper.find(".pre-use-legend").length) {
            report.page.wrapper.prepend(`
                <div class="pre-use-legend">
                    <div class="pre-use-legend-title">Pre-Use Report Legend</div>
                    <div>🟠 End Hours = 0</div>
                    <div>🔴 Working Hours = 0</div>
                    <div>🔴 Working Hours > 24</div>
                    <div>🟢 Normal Working Hours</div>
                </div>
            `);
        }
    },

    formatter: function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (!data) return value;

        if (column.fieldname === "asset_category") {
            return `<span style="font-weight:800;color:#102a43;">${value}</span>`;
        }

        if (column.fieldname === "asset_name") {
            return `<span class="pre-use-machine">🚜 ${value}</span>`;
        }

        if (column.fieldname === "item_name") {
            return `<span class="pre-use-muted">${value || ""}</span>`;
        }

        if (column.fieldname === "start_hours") {
            return `<span class="pre-use-number">${value}</span>`;
        }

        if (column.fieldname === "end_hours") {
            if (Number(data.end_hours) === 0) {
                return `<span class="pre-use-pill pre-use-pill-orange">${value}</span>`;
            }
            return `<span class="pre-use-number">${value}</span>`;
        }

        if (column.fieldname === "working_hours") {
            if (Number(data.working_hours) === 0) {
                return `<span class="pre-use-pill pre-use-pill-red">${value}</span>`;
            }

            if (Number(data.working_hours) > 24) {
                return `<span class="pre-use-pill pre-use-pill-red">${value}</span>`;
            }

            if (data.working_hours !== null && data.working_hours !== undefined) {
                return `<span class="pre-use-pill pre-use-pill-green">${value}</span>`;
            }
        }

        return value;
    }
};